mmengine/visualization/vis_backend.py

```diff
@@ -465,10 +478,13 @@ def add_scalar(self,
         Args:
             name (str): The scalar identifier.
             value (int, float, torch.Tensor, np.ndarray): Value to save.
-            step (int): Useless parameter. Wandb does not
-                need this parameter. Defaults to 0.
+            step (int): Global step value to record. Defaults to 0.
+                This will be ignored if auto_step is True.
         """
-        self._wandb.log({name: value}, commit=self._commit)
+        if self._auto_step:
+            self._wandb.log({name: value}, commit=self._commit)
+        else:
+            self._wandb.log({name: value}, commit=self._commit, step=step)
```

Comment:
@zhouzaida  thank you for your review.

To make it work, we need to change the default value of the `log_metric_by_epoch` in the `LoggerHook` as described in the "Note" section in this docstring.
I am not sure I can change the default value like `log_metric_by_epoch` in this PR.

The followings are the reason why `log_metric_by_epoch=False` is needed.

First, the `wandb.log` only supports monotonically increasing step sequences.
(See, https://docs.wandb.ai/guides/track/log/logging-faqs#why-am-i-seeing-fewer-data-points-than-i-logged )
If we set an earlier step in `wandb.log`, the `wandb` looks to ignore the log.
In the `LoggerHook`, the `runner.iter` is used as the step, but there is an exception, the `after_val_epoch` uses `runner.epoch` by default. 
Since almost always the `runner.epoch < runner.iter` is satisfied, we need to set `log_metric_by_epoch=False` to avoid the `wandb` ignoring the validation metrics.


One idea to remove the `auto_step` flag is to add a new attribute, `last_step`, in the `WandbVisBackend`. 
When the step < self.last_step is satisfied (this is probably runner.epoch), replace it with the `last_step`.

```py
if step < self.last_step:
    step = self.last_step
else:
    self.last_step = step
self._wandb.log(..., step=step)
```

I think this can ensure the step is always monotonically increasing. This may be more robust and better than `auto_flag` approach because we do not need to care about the value of `log_metric_by_epoch`.
Should I try this version?

Reply:
Hi @i-aki-y , thanks for your very clear explanation and I got your point. Due to the feature of wandb.log, we should increase the `step` monotonically. The idea you proposed looks good to me.